api/src/services/users/login_gov_callback_handler.py

```diff
@@ -0,0 +1,76 @@
+import logging
+from dataclasses import dataclass
+
+from pydantic import BaseModel
+from sqlalchemy import select
+
+import src.adapters.db as db
+from src.api.route_utils import raise_flask_error
+from src.db.models.user_models import LoginGovState
+from src.util.string_utils import is_valid_uuid
+
+logger = logging.getLogger(__name__)
+
+
+class CallbackParams(BaseModel):
+    code: str
+    state: str
+    error: str | None = None
+    error_description: str | None = None
+
+
+@dataclass
+class LoginGovCallbackResponse:
+    token: str
+    is_user_new: bool
+
+
+def handle_login_gov_callback(query_data: dict, db_session: db.Session) -> LoginGovCallbackResponse:
+    """Handle the callback from login.gov after calling the authenticate endpoint
+
+    NOTE: Any errors thrown here will actually lead to a redirect due to the
+          with_login_redirect_error_handler handler we have attached to the route
+    """
+
+    # Process the data coming back from login.gov via the redirect query params
+    # see: https://developers.login.gov/oidc/authorization/#authorization-response
+    callback_params = CallbackParams.model_validate(query_data)
+
+    # If we got an error back in the callback, raise an exception
+    # The only two documented error values are access_denied and invalid_request
+    # which would both indicate an issue in our configuration and we'll treat as a 5xx internal error
+    if callback_params.error is not None:
+        error_message = f"{callback_params.error} {callback_params.error_description}"
+        raise_flask_error(500, error_message)
+
+    # If the state value we received isn't a valid UUID
+    # then it's likely someone randomly calling the endpoint
+    # We don't want this validation on the schema as it would
+    # occur before our error catching that handles redirects
+    if not is_valid_uuid(callback_params.state):
+        raise_flask_error(422, "Invalid OAuth state value")
+
+    login_gov_state = db_session.execute(
+        select(LoginGovState).where(LoginGovState.login_gov_state_id == callback_params.state)
+    ).scalar_one_or_none()
+
+    # If we don't have the state value in our DB, that either means:
+    # * login.gov is very broken and sending us bad data
+    # * Someone called this endpoint directly with a random value
+    #
+    # There isn't a way to truly separate those here, so we'll assume the more likely second one
+    # and raise a 404 to say we have no idea what they passed us.
+    if login_gov_state is None:
```

Comment:
Are these secrets? Can they not be re-used? Logging what was passed could help us troubleshoot between the two scenarios you're anticipating for lookups failing if we don't consider them dangerous to log.

Reply:
Secret, probably not. Reused, definitely no.

State is basically the way we connect the initial call to login.gov with the callback. If it is none, that means someone probably called this endpoint directly with some random value and we don't have that ID in our system.

The nonce value stored is kinda the same idea, but instead it gets returned when we call the token endpoint (a later PR).

See: https://developers.login.gov/oidc/authorization/ for some further details.

Reply:
Cool, so I think logging the raw value which the error will indicate we didn't find will help us understand/troubleshoot any failed connectivity stuff we encounter here.